Merge MJIT infrastructure with conservative JIT compiler

k0kubun opened the pull request:
https://bugs.ruby-lang.org/issues/14235

## Background

In [Feature#12589](https://bugs.ruby-lang.org/issues/12589), Vladimir Makarov proposed to improve VM performance by replacing VM instructions
to RTL and introduce method JIT compiler based on those instructions.

While his approach for JIT (write C code to local file system, let C compiler executable
to compile it to shared object file and load it dynamically) was great and proven to work,
replacing all VM instructions may change the behavior of all Ruby programs and we can't turn off
such changes once it's released, even if we find a problem. So it's a little risky unlike optional JIT enablement.

Then I developed a JIT compiler called YARV-MJIT, which does not require any VM instruction changes.
After it, I heard Vladimir started to work on another approach to compile from current YARV instructions and
use RTL as IR for JIT compilation, but it's not published yet as far as I know.

## Problems

* We're developing the same JIT infrastructure independently, which can be shared for both implementations
  * it's definitely a waste of time, unlike seeking different optimization approaches
* If we continue to develop JIT in a big feature branch,
  * affected places will be big too, and thus it'll be a dangerous release
  * all of us will continue to waste our time by day-to-day conflict resolution against trunk
  * many valuable commit logs will be lost when we maintain the branch for rebase or squash commits on merge

## Solution

* **Proposal:** Merge MJIT infrastructure from Vladimir's patch with a conservative JIT compiler in early 2.6 development.
  * MJIT infrastructure means: JIT worker thread, profiler, gcc/clang compiler support, loading function from shared object file, some hooks to ensure JIT does not cause SEGV, etc...

### What's the "conservative JIT compiler"?

* Based on my YARV-MJIT, but this drops some problematic optimizations and is slower
* Pass `make test`, `make test-all`, `make test-spec` with and without JIT https://travis-ci.org/ruby/ruby/builds/321589821
* Unlike MJIT on RTL, we can play optcarrot (not just for benchmark, but on GUI) and run Rails application stably (not tested on production yet though)

### Notes

* As YARV-MJIT implementation improved MJIT infrastructure too, pthread was already ported to Windows native threads and it can be compiled with Visual Studio.
  * That's exactly why we should develop this in cooperation
* Visual Sudio is not supported as C compiler for JIT compilation. I did some experiments and had some ideas to support cl.exe, but I didn't want to add extra complexity to initial merge.
  * But it's perfectly working on MinGW and this will be available on Windows if a user uses RubyInstaller2.

### Optcarrot

Benchmarked with: Intel 4.0GHz i7-4790K with 16GB memory under x86-64 Ubuntu 8 Cores

|              |2.0.0      |2.5.0      |JIT off    |JIT on     |
|:-------------|:----------|:----------|:----------|:----------|
|optcarrot fps |35.05      |46.75      |46.05      |63.06      |
|vs 2.0.0      |1.00x      |1.33x      |1.31x      |1.80x      |

* 2.0.0: Ruby 2.0.0-p0
* 2.5.0: Ruby 2.5.0 (r61468)
* JIT off: Patched Ruby (based on r61475), JIT disabled
* JIT on: Patched Ruby (based on r61475), JIT enabled w/ gcc 5.4.0


Disclaimer: This JIT compiler performs better with gcc compared to clang for now, so it may be slow on macOS (clang).

### Micro benchmarks

I used Vladimir's benchmark set which I modified for my convenience https://github.com/benchmark-driver/mjit-benchmarks.

|           |2.0.0-p0   |2.5.0      |JIT off     |JIT on      |
|:----------|:----------|:----------|:----------|:----------|
|aread      |1.00       |1.01       |0.97       |2.33       |
|aref       |1.00       |0.96       |0.96       |3.01       |
|aset       |1.00       |1.39       |1.37       |3.70       |
|awrite     |1.00       |1.07       |1.03       |2.54       |
|call       |1.00       |1.25       |1.22       |3.39       |
|const      |1.00       |0.96       |0.96       |4.00       |
|const2     |1.00       |0.96       |0.96       |3.97       |
|fannk      |1.00       |0.98       |1.02       |1.00       |
|fib        |1.00       |1.16       |1.24       |3.19       |
|ivread     |1.00       |0.94       |0.93       |4.96       |
|ivwrite    |1.00       |1.09       |1.09       |3.32       |
|mandelbrot |1.00       |0.98       |0.98       |1.27       |
|meteor     |1.00       |3.02       |2.85       |3.16       |
|nbody      |1.00       |1.02       |0.99       |1.47       |
|nest-ntimes|1.00       |1.05       |1.01       |1.31       |
|nest-while |1.00       |0.96       |0.96       |1.63       |
|norm       |1.00       |1.06       |1.07       |1.26       |
|nsvb       |1.00       |0.98       |0.88       |0.88       |
|red-black  |1.00       |1.03       |1.02       |1.54       |
|sieve      |1.00       |1.22       |1.22       |1.75       |
|trees      |1.00       |1.07       |1.08       |1.32       |
|while      |1.00       |0.96       |0.96       |5.13       |

## Interface details

If the proposal is accepted, I'm going to add following CLI options:

```
  -j, --jit       use MJIT with default options
  -j:option, --jit:option
                  use MJIT with an option

MJIT options:
  c, cc           C compiler to generate native code (gcc, clang)
  s, save-temps   Save MJIT temporary files in $TMP or /tmp
  w, warnings     Enable printing MJIT warnings
  d, debug        Enable MJIT debugging (very slow)
  v=num, verbose=num
                  Print MJIT logs of level num or less to stderr
  n=num, num-cache=num
                  Maximum number of JIT codes in a cache
```

Note that `-j:l`/`--jit:llvm` are changed to `-j:c`/`--jit:cc` so that we can support cl.exe (Visual Studio) in the future.

Also, for testing, I would like to have following module and method.

```rb
MJIT.enabled? #=> true / false
```

See the commit log for details.

nobu commented on mjit.h:
```diff
@@ -0,0 +1,125 @@
+
+extern int mjit_compile(FILE *f, const struct rb_iseq_constant_body *body, const char *funcname);
+extern void mjit_init(struct mjit_options *opts);
+extern void mjit_finish();
```
Some functions are declared/defined in old K&R style.

k0kubun replied:
Unified the style in https://github.com/ruby/ruby/pull/1782/commits/e22a8e263383b6f924b94d29f65b677344b5daa9

nurse commented on thread_pthread.c:
```diff
@@ -1788,4 +1789,36 @@ rb_nativethread_self(void)
+
+    if (pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL) != 0) {
+	fprintf(stderr, "Cannot enable cancelation in MJIT worker\n");
+    }
```
It's nice to set thread name:
```c
#ifdef SET_CURRENT_THREAD_NAME
    SET_CURRENT_THREAD_NAME("ruby-mjitworker");
#endif
```
Note that the max size of the name is 16 byte including NUL byte of the string.
A natural name of the thread "ruby-mjit-worker" is 16 byte; it exceeds 1 byte.

k0kubun replied:
Thank you to catch, fixed https://github.com/ruby/ruby/pull/1782/commits/ddb2ff7303982b19adb0a23707722e3ca4ece1d2

nurse commented on thread_win32.c:
```diff
@@ -777,4 +777,26 @@ native_set_thread_name(rb_thread_t *th)
+static unsigned long __stdcall
+mjit_worker(void *arg)
+{
+    void (*worker_func)(void) = arg;
```
Add following:
```c
rb_w32_set_thread_description(GetCurrentThread(), L"ruby-mjitworker");
```
[SetThreadDescription](https://msdn.microsoft.com/en-us/library/windows/desktop/mt774976(v=vs.85).aspx) doesn't have the max size limit but I think the name should equal to Unix.

k0kubun replied:
fixed at https://github.com/ruby/ruby/pull/1782/commits/ddb2ff7303982b19adb0a23707722e3ca4ece1d2

rurban commented on ruby.c:
```diff
@@ -272,6 +272,7 @@ usage(const char *name, int help)
 	M("s",     ", save-temps",     "Save MJIT temporary files in $TMP or /tmp"),
 	M("w",     ", warnings",       "Enable printing MJIT warnings"),
 	M("d",     ", debug",          "Enable MJIT debugging (very slow)"),
+	M("a=num", ", aot=num",        "Ahead of Time Compileation after num calls"),
```
typo: Compileation

k0kubun replied:
Thank you https://github.com/ruby/ruby/pull/1782/commits/a87c8ef5319a020aeeca241f4ac05dbc60c89367

radarek commented:
Where I should report issues related to mjit? I did some performance tests and I would like to share my results and observations with you.

Fudoshiki commented:
@radarek 
https://github.com/k0kubun/ruby
https://github.com/k0kubun/yarv-mjit

k0kubun commented:
Thank you for your comments. Final results should be written to https://bugs.ruby-lang.org/issues/14235 for easy tracking, but at this moment here or https://github.com/k0kubun/yarv-mjit/issues is better.

That's because the result may be changed before merge. To remove all bugs, I'm going to add some guards and the benchmark score will be a little worse after that.

k0kubun commented:
I added some guards to let all tests pass with AOT compilation, and removed some optimizations to omit SEGV by race condition (which should be fixed and added later). 

Then unfortunately performance is made worse as follows :disappointed:, but now it's probably ready to merge.

| | with JIT ([before fix](https://github.com/ruby/ruby/pull/1782/commits/db6e3e4f2759f17b302805f8d193b3cc1c2a5f74)) | with JIT ([after fix](https://github.com/ruby/ruby/pull/1782/commits/7efca52e7b24fca33b812eeef32aa0403f65ac1f)) | without JIT |
|:---|:-------|:-------|:------|
| optcarrot fps | 60.96 | 52.72 | 45.50 |

I'll add other revised benchmarks too.

k0kubun commented:
## Current status
I got some ideas to resurrect above performance to original proposal's one. 

Until Jan 24th (next Ruby developers meeting at Tokyo), I'll work on it, try to figure out why it becomes worse with Rails, and create a script that generates `compile_insn` from insns.def for maintainability. 

If things go well, probably this (or part of this) will be merged on that day.

k0kubun commented:
While current trunk's performance is worse than my original benchmark and I have some uncommitted ideas to fix it, most of this work is merged in https://github.com/ruby/ruby/commit/0af44e72611583963220e40215fc36f31fa80185, https://github.com/ruby/ruby/commit/fd44a5777fbca386a78cccff9aa02eca155b0752 and https://github.com/ruby/ruby/commit/ed935aa5be0e5e6b8d53c3e7d76a9ce395dfa18b.

There are still a lot of things to be done, but it's time to close this PR.
